lib/source/mqtt/aws_iot_mqtt_api.c

```diff
@@ -1696,110 +1738,150 @@ AwsIotMqttError_t AwsIotMqtt_Wait( AwsIotMqttReference_t reference,
     uint64_t startTime = 0, currentTime = 0, elapsedTime = 0, remainingMs = timeoutMs;
     AwsIotMqttError_t status = AWS_IOT_MQTT_STATUS_PENDING;
     _mqttOperation_t * pOperation = ( _mqttOperation_t * ) reference;
+    _mqttConnection_t * pMqttConnection = pOperation->pMqttConnection;
 
     /* Check reference. */
     if( AwsIotMqttInternal_ValidateReference( reference ) == false )
     {
         return AWS_IOT_MQTT_BAD_PARAMETER;
     }
 
-    IotLogInfo( "Waiting for operation %s to complete.",
-                AwsIotMqtt_OperationType( pOperation->operation ) );
+    /* Lock the connection mutex to prevent the connection from being closed. */
+    IotMutex_Lock( &( pMqttConnection->connectionMutex ) );
 
-    /* Wait for the operation to be sent once. This wait should return quickly. */
-    IotSemaphore_Wait( &( pOperation->notify.waitSemaphore ) );
+    /* Check connection status. */
+    if( ( pOperation->operation != AWS_IOT_MQTT_DISCONNECT ) &&
+        ( pMqttConnection->disconnected == true ) )
+    {
+        /* Return "Send Error" for a closed connection. */
+        status = AWS_IOT_MQTT_SEND_ERROR;
+    }
+    else
+    {
+        IotLogInfo( "Waiting for operation %s to complete.",
+                    AwsIotMqtt_OperationType( pOperation->operation ) );
 
-    /* Check any status set by the send thread. Block the receive callback
-     * during this check by locking the mutex for operations pending responses. */
-    IotMutex_Lock( &( _IotMqttPendingResponseMutex ) );
-    status = pOperation->status;
-    IotMutex_Unlock( &( _IotMqttPendingResponseMutex ) );
+        /* Wait for the operation to be sent once. This wait should return quickly. */
+        IotSemaphore_Wait( &( pOperation->notify.waitSemaphore ) );
+    }
+
+    IotMutex_Unlock( &( pMqttConnection->connectionMutex ) );
 
     if( status == AWS_IOT_MQTT_STATUS_PENDING )
     {
-        /* Check if this operation is a PUBLISH with retry. Block the timer
-         * thread during this check by locking the connection mutex. */
-        if( pOperation->operation == AWS_IOT_MQTT_PUBLISH_TO_SERVER )
-        {
-            IotMutex_Lock( &( pOperation->pMqttConnection->timerMutex ) );
-            publishRetryActive = ( pOperation->pPublishRetry != NULL );
-            IotMutex_Unlock( &( pOperation->pMqttConnection->timerMutex ) );
-        }
+        /* Check any status set by the send thread. Block the receive callback
+         * during this check by locking the mutex for operations pending responses. */
+        IotMutex_Lock( &( _IotMqttPendingResponseMutex ) );
+        status = pOperation->status;
+        IotMutex_Unlock( &( _IotMqttPendingResponseMutex ) );
```

Comment:
After you release the mutex, your status variable can be changed, and there s really no reason therefore to read it under a lock.
Different would be if you read it AND update it under lock, so that any thread that comes after will take a different path. But that is not the case here...